tokio-util/src/io/sink_writer.rs

```diff
@@ -0,0 +1,183 @@
+    /// A helper structure, converting a sink of byte slices into a
+    /// sink of owned [`Bytes`].
+    ///
+    ///
+    /// [`Bytes`]: bytes::Bytes
+    #[derive(Debug)]
+    pub struct CopyToBytes<S> {
+        #[pin]
+        inner: S,
+    }
+}
+
+impl<S> CopyToBytes<S> {
+    /// Creates a new [`CopyToBytes<S>`].
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+
+    /// Gets a reference to the underlying sink.
+    ///
+    /// It is inadvisable to directly write to the underlying sink.
+    pub fn get_ref(&self) -> &S {
+        &self.inner
+    }
+
+    /// Gets a mutable reference to the underlying sink.
+    ///
+    /// It is inadvisable to directly write to the underlying sink.
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.inner
+    }
+
+    /// Consumes this `SinkWriter`, returning the underlying sink.
+    ///
+    /// It is inadvisable to directly write to the underlying sink.
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
+}
+
+impl<'a, S> Sink<&'a [u8]> for CopyToBytes<S>
+where
+    S: Sink<Bytes>,
+{
+    type Error = S::Error;
+
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.as_mut().project().inner.poll_ready(cx)
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: &'a [u8]) -> Result<(), Self::Error> {
+        self.as_mut()
+            .project()
+            .inner
+            .start_send(Bytes::copy_from_slice(item))
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.as_mut().project().inner.poll_flush(cx)
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.as_mut().project().inner.poll_close(cx)
+    }
+}
+
+impl<S, E> AsyncWrite for SinkWriter<S>
+where
+    for<'a> S: Sink<&'a [u8], Error = E>,
+    E: Into<io::Error>,
+{
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        match self.as_mut().project().inner.poll_ready(cx) {
+            Poll::Ready(Ok(())) => {
+                if let Err(e) = self.as_mut().project().inner.start_send(buf) {
+                    Poll::Ready(Err(e.into()))
+                } else {
+                    Poll::Ready(Ok(buf.len()))
+                }
+            }
+            Poll::Ready(Err(e)) => Poll::Ready(Err(e.into())),
+            Poll::Pending => {
+                cx.waker().wake_by_ref();
+                Poll::Pending
+            }
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        self.project().inner.poll_flush(cx).map_err(Into::into)
+    }
+
+    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        self.project().inner.poll_close(cx).map_err(Into::into)
+    }
+}
```

Comment:
Minor formatting nit - there's no reason to separate this `impl` from the rest of `SinkWriter`, as it's not affected by `CopyToBytes` existing. Can it be moved to sit with the other `SinkWriter` `impl`s? Either move them down here, or move this one to be with the others.

Reply:
@Darksonn had the proposal to move `CopyToBytes` into a separate file. I assume, given the primary intended use case, it is still a good idea to keep that in `io`.